frontend/src/metabase-lib/Question.ts

```diff
@@ -1286,6 +1298,211 @@ class QuestionInner {
     return hasQueryBeenAltered ? question.markDirty() : question;
   }
 
+  generateQueryDescription(tableMetadata, options = {}) {
+    if (!tableMetadata || (this.isNative() && !this.displayName())) {
+      return "";
+    }
+
+    options = {
+      sections: [
+        "table",
+        "aggregation",
+        "breakout",
+        "filter",
+        "order-by",
+        "limit",
+      ],
+      ...options,
+    };
+
+    const sectionFns = {
+      table: this._getTableDescription.bind(this),
+      aggregation: this._getAggregationDescription.bind(this),
+      breakout: this._getBreakoutDescription.bind(this),
+      filter: this._getFilterDescription.bind(this),
+      "order-by": this._getOrderByDescription.bind(this),
+      limit: this._getLimitDescription.bind(this),
+    };
+
+    // these array gymnastics are needed to support JSX formatting
+    const query = this.datasetQuery().query;
+    const sections = options.sections
+      .map(section =>
+        _.flatten(sectionFns[section](tableMetadata, query, options)).filter(
+          s => !!s,
+        ),
+      )
+      .filter(s => s && s.length > 0);
+
+    const description = _.flatten(DESCRIPTION.joinList(sections, ", "));
+    return description.join("");
```

Comment:
FWIW, there is an `Array.prototype.flat` method we can use instead of `_.flatten`